File: 396 rotate-function/js/main.js
/**
 * @param {number[]} A
 * @return {number}
 */

// F(i + 1) = F(i) + SUM - A[n - i] * n

var maxRotateFunction = function(A) {
  let len = A.length
  if (len <= 1) return 0

  let max = -Infinity
  let sum = 0
  let fi = 0
  for (let i = 0; i < len; i++) {
    sum += A[i]
    fi += A[i] * i
  }

  max = Math.max(max, fi)
  for (let i = 1; i < len; i++) {
    fi = fi + sum - len * A[len - i]
    max = Math.max(max, fi)
  }

  return max
};

function test(arr) {
  console.log('max is', maxRotateFunction(arr))
}

test([])
test([1])
test([4, 3, 2, 6])

File: 396 rotate-function/c/main.c
#include <limits.h>

#define MAX(a, b) (a > b ? a : b)

int maxRotateFunction(int* A, int ASize) {
  if (ASize <= 1)
    return 0;

  int sum = 0;
  int fi = 0;
  for (int i = 0; i < ASize; i++) {
    sum += A[i];
    fi += A[i] * i;
  }

  int max = fi;
  for (int i = 1; i < ASize; i++) {
    fi = fi + sum - ASize * A[ASize - i];
    max = MAX(max, fi);
  }

  return max;
}

int main(void) {
  return 0;
}
